task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/GiftCertificateRepositoryImpl.java

```diff
@@ -0,0 +1,321 @@
+package com.epam.esm.repository.impl;
+
+import com.epam.esm.entity.GiftCertificate;
+import com.epam.esm.entity.Tag;
+import com.epam.esm.repository.GiftCertificateRepository;
+import com.epam.esm.repository.util.QueryUtil;
+import com.epam.esm.repository.exception.RepositoryException;
+import com.epam.esm.repository.impl.mapper.GiftCertificateMapper;
+import com.epam.esm.repository.impl.mapper.TagMapper;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DataAccessException;
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.data.domain.Sort;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.PreparedStatementCreator;
+import org.springframework.jdbc.core.ResultSetExtractor;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+
+import java.sql.PreparedStatement;
+import java.sql.Timestamp;
+import java.util.*;
+
+@Repository
+public class GiftCertificateRepositoryImpl implements GiftCertificateRepository {
+    private static final Logger logger = LogManager.getLogger(GiftCertificateRepositoryImpl.class);
+
+    @Autowired
+    private JdbcTemplate jdbcTemplate;
+
+    private static final String FIND_ALL_QUERY =
+            "SELECT id gc_id, name gc_name, description, price, duration, create_date, last_update_date " +
+            "FROM gift_certificate";
+    private static final String FIND_ALL_WITH_TAGS_QUERY =
```

Comment:
@igor-zhuravlev A lot of copypast.  Your queries are almost the same in terms of parameters and differ only in search and sorting parameters. Why not generate all this in one place?